File: Example-objc/AcaiaScaleTableVC.m
//
//  AcaiaScaleTableVC.m
//  AcaiaExample
//
//  Created by Michael Wu on 2018/7/2.
//  Copyright © 2018 mikewu1211. All rights reserved.
//

#import "AcaiaScaleTableVC.h"
#import "AppDelegate.h"
@import AcaiaSDK;

@interface AcaiaScaleTableVC ()
@property BOOL isRefreshed;
@property NSTimer *timer;
@property UIActivityIndicatorView* activityIndicatorView;
@end

@implementation AcaiaScaleTableVC

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.isRefreshed = NO;
    
    self.refreshControl = [[UIRefreshControl alloc] init];
    [self.refreshControl addTarget:self
                            action:@selector(refresh:)
                  forControlEvents:UIControlEventValueChanged];
    self.refreshControl.attributedTitle = [[NSAttributedString alloc] initWithString:@"Scanning"];
    self.tableView.refreshControl = self.refreshControl;
    
    self.activityIndicatorView = [[UIActivityIndicatorView alloc]
                                  initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleLarge];
    self.activityIndicatorView.hidesWhenStopped = true;
    self.activityIndicatorView.hidden = true;
    [self.view addSubview:self.activityIndicatorView];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(onConnect:)
                                                 name:AcaiaScaleDidConnected
                                               object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(onFinishScan:)
                                                 name:AcaiaScaleDidFinishScan
                                               object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(scanListChanged:)
                                                 name:AcaiaScaleDeviceListChanged
                                               object:nil];
}

- (void)viewWillAppear:(BOOL)animated {
    self.activityIndicatorView.center = self.view.center;
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    
    [[AcaiaManager sharedManager] startScan:0.5];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)refresh:(UIRefreshControl *)refreshCtrl {
    [[AcaiaManager sharedManager] startScan:0.5];
}

#pragma mark - Acaia SDK Notification

- (void)onFinishScan:(NSNotification *)noti {
    [self.tableView.refreshControl endRefreshing];

    if ([AcaiaManager sharedManager].scaleList.count > 0) {
        [self.tableView reloadData];
    }
}

- (void) scanListChanged:(NSNotification *)noti {
    [self.tableView.refreshControl endRefreshing];
    [self.tableView reloadData];
}

- (void)onConnect:(NSNotification *)noti {
    [self.activityIndicatorView stopAnimating];
    [self.navigationController popViewControllerAnimated:YES];
    if (self.timer != nil) {
        [self.timer invalidate];
        self.timer = nil;
    }
}

- (void)onTimer:(NSTimer *)timer {
    [self.activityIndicatorView stopAnimating];
    [self.timer invalidate];
    self.timer = nil;
    
    [AcaiaManager.sharedManager startScan:0.1];
    
    UIAlertController* alert =
    [UIAlertController alertControllerWithTitle:nil
                                        message:@"Connect timeout"
                                 preferredStyle:UIAlertControllerStyleAlert];
    
    [alert addAction:[UIAlertAction actionWithTitle:@"Ok"
                                              style:UIAlertActionStyleCancel
                                            handler:nil]];
    
    [self presentViewController:alert animated:true completion:nil];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [AcaiaManager sharedManager].scaleList.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"ScaleCell"
                                                            forIndexPath:indexPath];
    
    AcaiaScale* scale = [AcaiaManager sharedManager].scaleList[indexPath.row];
    
    cell.textLabel.text = scale.name;
    cell.detailTextLabel.text = scale.modelName;
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    self.timer  = [NSTimer scheduledTimerWithTimeInterval:10.0f
                                                   target:self
                                                 selector:@selector(onTimer:)
                                                 userInfo:nil
                                                  repeats:NO];
    AcaiaScale* scale = [AcaiaManager sharedManager].scaleList[indexPath.row];
    [scale connect];
    
    [self.activityIndicatorView startAnimating];
}

@end

File: Example-swift/AcaiaScaleTableVC.swift
//
//  AcaiaScaleTableVC.swift
//  AcaiaSDKExampleSwift
//
//  Created by Michael Wu on 2018/7/4.
//  Copyright © 2018 acaia Corp. All rights reserved.
//

import UIKit
import AcaiaSDK


class AcaiaScaleTableVC: UITableViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        
        _setRefreshControl()
        _addAcaiaEventsObserver()
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        AcaiaManager.shared().startScan(0.5)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        _activityIndicatorView.stopAnimating()
        
        _timerForConnectTimeOut?.invalidate()
        _timerForConnectTimeOut = nil
        
        _removeScaleEventObservers()
    }

    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return AcaiaManager.shared().scaleList.count
    }

    override func tableView(_ tableView: UITableView,
                            cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "ScaleCell", for: indexPath)

        let scale = AcaiaManager.shared().scaleList[indexPath.row]

        cell.textLabel?.text = scale.name
        cell.detailTextLabel?.text = scale.modelName
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        _activityIndicatorView.isHidden = false
        _activityIndicatorView.startAnimating()
        
        let scale = AcaiaManager.shared().scaleList[indexPath.row]
        scale.connect()
        
        _timerForConnectTimeOut = Timer.scheduledTimer(timeInterval: 10.0,
                                                       target: self,
                                                       selector: #selector(_connectTimeOut(_:)),
                                                       userInfo: nil,
                                                       repeats: false)
    }
    
    
    // MARK: Private
    
    private var _timerForConnectTimeOut: Timer? = nil
    
    lazy private var _activityIndicatorView: UIActivityIndicatorView = {
        let indicator = UIActivityIndicatorView(style: .large)
        indicator.hidesWhenStopped = true
        indicator.isHidden = true
        
        view.addSubview(indicator)
        indicator.center = view.center
        
        return indicator
    }()
    
    private func _setRefreshControl() {
        tableView.refreshControl = .init()
        tableView.refreshControl?.attributedTitle = .init(string: "Scanning")
        tableView.refreshControl?.addAction(
            UIAction { _ in AcaiaManager.shared().startScan(0.5) },
            for: .valueChanged
        )
    }
    
    private func _addAcaiaEventsObserver() {
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(_didConnected),
                                               name: .init(rawValue: AcaiaScaleDidConnected),
                                               object: nil)
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(_didFinishScan),
                                               name: .init(rawValue: AcaiaScaleDidFinishScan),
                                               object: nil)
    }
    
    private func _removeScaleEventObservers() {
        NotificationCenter.default.removeObserver(self)
    }
    
    @objc private func _didConnected(notification: NSNotification) {
        _activityIndicatorView.stopAnimating()
        
        _timerForConnectTimeOut?.invalidate()
        _timerForConnectTimeOut = nil
        
        navigationController?.popViewController(animated: true)
    }
    
    @objc private func _didFinishScan(notification: NSNotification) {
        tableView.refreshControl?.endRefreshing()
        tableView.reloadData()
    }
    
    @objc private func _connectTimeOut(_ timer: Timer) {
        _activityIndicatorView.stopAnimating()
        
        _timerForConnectTimeOut?.invalidate()
        _timerForConnectTimeOut = nil
        
        AcaiaManager.shared().startScan(0.1)
        
        let alert = UIAlertController(title: nil,
                                      message: "Connect timeout",
                                      preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "Ok", style: .cancel))
        
        present(alert, animated: true)
    }
}

